app/src/endpoints.c

```diff
@@ -16,44 +16,90 @@
-#include <zmk/events/endpoint_selection_changed.h>
+#include <zmk/events/endpoint_changed.h>
 
 #include <zephyr/logging/log.h>
 LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
-#define DEFAULT_ENDPOINT                                                                           \
-    COND_CODE_1(IS_ENABLED(CONFIG_ZMK_BLE), (ZMK_ENDPOINT_BLE), (ZMK_ENDPOINT_USB))
+#define DEFAULT_TRANSPORT                                                                          \
+    COND_CODE_1(IS_ENABLED(CONFIG_ZMK_BLE), (ZMK_TRANSPORT_BLE), (ZMK_TRANSPORT_USB))
 
-static enum zmk_endpoint current_endpoint = DEFAULT_ENDPOINT;
-static enum zmk_endpoint preferred_endpoint =
-    ZMK_ENDPOINT_USB; /* Used if multiple endpoints are ready */
+static struct zmk_endpoint_instance current_instance = {};
+static enum zmk_transport preferred_transport =
+    ZMK_TRANSPORT_USB; /* Used if multiple endpoints are ready */
 
-static void update_current_endpoint();
+static void update_current_endpoint(void);
 
 #if IS_ENABLED(CONFIG_SETTINGS)
 static void endpoints_save_preferred_work(struct k_work *work) {
-    settings_save_one("endpoints/preferred", &preferred_endpoint, sizeof(preferred_endpoint));
+    settings_save_one("endpoints/preferred", &preferred_transport, sizeof(preferred_transport));
 }
 
 static struct k_work_delayable endpoints_save_work;
 #endif
 
-static int endpoints_save_preferred() {
+static int endpoints_save_preferred(void) {
 #if IS_ENABLED(CONFIG_SETTINGS)
     return k_work_reschedule(&endpoints_save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
 #else
     return 0;
 #endif
 }
 
-int zmk_endpoints_select(enum zmk_endpoint endpoint) {
-    LOG_DBG("Selected endpoint %d", endpoint);
+bool zmk_endpoint_instance_equals(struct zmk_endpoint_instance a, struct zmk_endpoint_instance b) {
+    if (a.transport != b.transport) {
+        return false;
+    }
+
+    switch (a.transport) {
+    case ZMK_TRANSPORT_USB:
+        return true;
+
+    case ZMK_TRANSPORT_BLE:
+        return a.ble_profile_index == b.ble_profile_index;
+    }
+
+    LOG_ERR("Invalid transport %d", a.transport);
+    return false;
+}
+
+int zmk_endpoint_instance_print(char *str, size_t len, struct zmk_endpoint_instance endpoint) {
+    switch (endpoint.transport) {
+    case ZMK_TRANSPORT_USB:
+        return snprintf(str, len, "USB");
+
+    case ZMK_TRANSPORT_BLE:
+        return snprintf(str, len, "BLE:%d", endpoint.ble_profile_index);
+
+    default:
+        return snprintf(str, len, "Invalid");
+    }
+}
+
+#define INSTANCE_INDEX_OFFSET_USB 0
+#define INSTANCE_INDEX_OFFSET_BLE ZMK_ENDPOINT_USB_COUNT
+
+int zmk_endpoint_instance_to_index(struct zmk_endpoint_instance endpoint) {
+    switch (endpoint.transport) {
+    case ZMK_TRANSPORT_USB:
+        return INSTANCE_INDEX_OFFSET_USB;
+
+    case ZMK_TRANSPORT_BLE:
+        return INSTANCE_INDEX_OFFSET_BLE + endpoint.ble_profile_index;
+    }
+
+    LOG_ERR("Invalid transport %d", endpoint.transport);
+    return 0;
```

Comment:
Does that make any invalid case return the same `0` value as the valid USB case?

Reply:
Yes. I didn't want to pick a value that wasn't a valid array index, because then everything that uses this would have to add error checking for a case that shouldn't ever occur.

Reply:
> Yes. I didn't want to pick a value that wasn't a valid array index, because then everything that uses this would have to add error checking 

I assumed that was the reason

> for a case that shouldn't ever occur.

My concern was silently overriding existing data when passing an invalid instance but if that invalid case is practically unreachable then nevermind

Reply:
Should the return type be unsigned if there's no error codes?

Reply:
I see, I was under the impression array indices were better as `uint` or `size_t` but it's apparently not that simple https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es107-dont-use-unsigned-for-subscripts-prefer-gslindex. Thanks for the detailed answer.